### ISSUE TITLE
Aborting launch before app is fully launched leaves the 'flutter run' process running.

### ISSUE THREAD

**Issue by knopp** [Dart-Code/Dart-Code#772]
When stopping right after launch, the dart process seems to be kept running instead being killed, which breaks subsequent attempts to launch app. These processes keep piling up and the app will fail to launch until all of them quit.

**Comment by DanTup** [comment]
I can't reproduce this no matter where I terminate. Could you let me know:

- At what point you're terminating
- What process(es) are hanging around when this happens
- What is the error you get for subsequent launches?

Additionally, the [Flutter Run](https://dartcode.org/docs/logging/#flutter-run) logs would be useful, though note that they get overwritten on each run - so after terminating, please take a copy of the log before attempting to run the next time (and then send both logs, from the early-terminated version, and the next one that fails to run).

Thanks!

**Comment by knopp** [comment]
Hi, I looked at the logs, I don't think they contain anything interesting. In any case, it seems to happen when I terminate at any point before the app is fully launched (i.e. building or installation).

When I do 
```ps x | grep dart```

I get something like this
```
62134   ??  S      0:03.36 /Users/Matej/Projects/flutter/flutter/bin/cache/dart-sdk/bin/dart /Users/Matej/Projects/flutter/flutter/bin/cache/flutter_tools.snapshot run --machine -t lib/main.dart -d f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c --preview-dart-2 --start-paused
62623   ??  S      0:00.03 ios-deploy --id f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c --bundle build/ios/iphoneos/Runner.app --no-wifi --justlaunch --args --enable-dart-profiling --start-paused --enable-checked-mode
```

Eventually those processes quit on their own, but long VSCode stop button is pressed. I am very curious as for why this doesn't happen for you

Also killing the process normally (SIGTERM) from command line seem to terminate the process without issues

Here's log from launch
```
[11:10:18 AM]: Spawning /Users/Matej/Projects/flutter/flutter/bin/flutter with args ["run","--machine","-t","lib/main.dart","-d","f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c","--preview-dart-2","--start-paused"]
[11:10:18 AM]: ..  in /Users/Matej/Projects/flutter/remote1
[11:10:18 AM]: ..  in {"FLUTTER_HOST":"VSCode"}
[11:10:20 AM]: <== [{"event":"daemon.showMessage","params":{"level":"warning","title":"Unable to list devices","message":"Unable to discover Android devices. Please run \"flutter doctor\" to diagnose potential issues"}}]
[11:10:20 AM]: <== [{"event":"app.start","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","deviceId":"f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c","directory":"/Users/Matej/Projects/flutter/remote1","supportsRestart":true}}]
[11:10:21 AM]: <== Launching lib/main.dart on Matej Knopp’s iPhone X in debug mode...
[11:10:29 AM]: <== Found saved certificate choice "iPhone Developer: Matej Knopp (QJW2VK38XZ)". To clear, use "flutter config".
[11:10:29 AM]: <== Signing iOS app for device deployment using developer identity: "iPhone Developer: Matej Knopp (QJW2VK38XZ)"
[11:10:29 AM]: <== [{"event":"app.progress","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","id":"0","progressId":null,"message":"Running Xcode clean..."}}]
[11:10:31 AM]: <== [{"event":"app.progress","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","id":"0","progressId":null,"finished":true}}]
[11:10:31 AM]: <== [{"event":"app.progress","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","id":"1","progressId":null,"message":"Starting Xcode build..."}}]
[11:10:41 AM]: ==> [{"id":"1","method":"app.stop","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108"}}]
[11:10:41 AM]: <== Application finished.
[11:10:41 AM]: <== [{"id":"1","result":true}]
[11:10:47 AM]: <== [{"event":"app.progress","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","id":"1","progressId":null,"finished":true}}]
[11:10:47 AM]: <== Xcode build done
[11:10:49 AM]: <== [{"event":"app.progress","params":{"appId":"3f6cc29a-1623-439f-864c-75c884dfa108","id":"2","progressId":null,"message":"Installing and launching..."}}]
```

**Comment by DanTup** [comment]
Hmmm, I think there might be a bug here in flutter. We can see an `app.stop` request being sent at `11:10:41` but it seems to carry on doing stuff.

@devoncarew Presumably we should be able to send a stop during the build and not have to wait for the app to launch?

**Comment by knopp** [comment]
Maybe as a fallback you could do sigterm after some delay if the app.stop doesn't quit it?

**Comment by DanTup** [comment]
Hmmm, based on the code I just saw when investigating #771 we are killing it too. I'll do some more digging and see what's going on.

Btw, can you repro this with the simulator?

**Comment by knopp** [comment]
I can reproduce this on Simulator. I stop the launch process right after starting, but the process still complete with app being launched in simulator (stuck with white screen because debugger doesn't attach)

**Comment by DanTup** [comment]
Ok, I've managed to repro this. I've made a change to wait for confirmation of the stop before `.kill()`ing, however it doesn't solve the issue. I also changed it to never try to kill, but wait for the process to end - it never does, so the debug toolbar/debug session never sends.

I've opened https://github.com/flutter/flutter/issues/16183 for investigation/comments. I'm not sure I can fix this in a reliable way from Dart Code.

**Comment by knopp** [comment]
Have you tried simply killing it without sending the quit command? SIGTERM from terminal seem to work quite reliably, I don't quite understand why sigterm from vscode wouldn't.

**Comment by DanTup** [comment]
Tried that, made no difference :( I think there's quite a difference in how things work with the `--machine` flag to the command line (there have been other issues that didn't occur from the command line but did occur from the API) so I'm not all that surprised by differences in behaviour.

I tried `flutter run --machine` from the command line and hitting `Ctrl+C` but that also seems fine (though I think this sends SIGINT instead of SIGTERM).

**Comment by DanTup** [comment]
Hmm, even sending a sigterm from the command line it seems to work fine. I'll keep digging (and adding more logging....)

**Comment by DanTup** [comment]
Random thought - we're spawning with `shell: true` to fix Windows issues, I wonder whether this changes things (the process we're sending sigterm to I guess is therefore the shell?) ...

**Comment by DanTup** [comment]
😢 
https://github.com/pkrumins/node-tree-kill/blob/master/index.js

**Comment by knopp** [comment]
I have no experience with nodejs's spawn, but I assume if it launches shell to run the child process and thus it sends the SIGTERM to that shell, which should forward it to child it spawned, so I don't understand why the kill-tree would be necessary.

**Comment by DanTup** [comment]
That's what I'd have expected too, but honestly I'm not surprised by any crazy behaviour in nodejs these days =)

I found that `daemon.shutdown` is available in `flutter run --machine` so tried sending that, but it also didn't work. I think any fix I apply is going to be a nasty hack (and potentially unreliable), so I'm gonna wait to heat back on https://github.com/flutter/flutter/issues/16183 as I think if stop/shutdown works reliably this problem should disappear.

**Comment by DanTup** [comment]
Right, I think I finally fixed this. I'm landing [a change](https://github.com/flutter/flutter/pull/17873) in Flutter that will expose the `pid` of the `flutter run` process, and we added that to the list of "additional processes to terminate".

I couldn't repro on  macOS today, but I did repro on Windows, and the fix seems to be reliable. It's a bit late for v2.13, but if you want an early build of v2.14 with this fix after this week (you'll need to also be on flutter master) give me a shout.
